mesonbuild/mesonlib.py

```diff
@@ -25,25 +25,33 @@
 
 def detect_meson_py_location():
     c = sys.argv[0]
-    c_fname = os.path.split(c)[1]
-    if c_fname == 'meson' or c_fname == 'meson.py':
-        # $ /foo/meson.py <args>
-        if os.path.isabs(c):
-            return c
-        # $ meson <args> (gets run from /usr/bin/meson)
+    c_dir, c_fname = os.path.split(c)
+
+    # get the absolute path to the <mesontool> folder
+    m_dir = None
+    if os.path.isabs(c):
+        # $ /foo/<mesontool>.py <args>
+        m_dir = c_dir
+    elif c_dir == '':
+        # $ <mesontool> <args> (gets run from /usr/bin/<mesontool>)
         in_path_exe = shutil.which(c_fname)
         if in_path_exe:
+            m_dir = os.path.dirname(in_path_exe)
             # Special case: when run like "./meson.py <opts>" and user has
             # period in PATH, we need to expand it out, because, for example,
             # "ninja test" will be run from a different directory.
-            if '.' in os.environ['PATH'].split(':'):
-                p, f = os.path.split(in_path_exe)
-                if p == '' or p == '.':
-                    return os.path.join(os.getcwd(), f)
-            return in_path_exe
-        # $ python3 ./meson.py <args>
-        if os.path.exists(c):
-            return os.path.join(os.getcwd(), c)
+            if ('.' in os.environ['PATH'].split(':') and
+                (m_dir == '' or m_dir == '.')):
+                m_dir = os.getcwd()
+    else:
+        m_dir = os.path.abspath(c_dir)
+
+    # find meson in m_dir
+    if not m_dir is None:
```

Comment:
__[Flake8]__

[E714] test for object identity should be 'is not'


<!-- Comment by [SideCI](https://sideci.com) -->